lib/ast.ex

```diff
@@ -0,0 +1,101 @@
+    cond do
+      Map.has_key?(map, :body) ->
+        %{map | body: remap_with_gamma(map.body)}
+      Map.has_key?(map, :value) ->
+        %{map | value: remap_with_gamma(map.value)}
+      true ->
+        map
+    end
+  end
+
+  def remap_with_gamma([computation | rest], new_ast \\ []) do
+    comp = remap_with_gamma(computation)
+
+    case comp do
+      %ESTree.MethodDefinition{} -> remap_with_gamma(rest, new_ast ++ [comp])
+      %ESTree.ClassDeclaration{} -> remap_with_gamma(rest, new_ast ++ [comp])
+      _ -> remap_with_gamma(rest, (new_ast ++ [generate_gamma_charge(comp), sanitize_computation(comp)]))
+    end
+  end
+
+  def remap_with_gamma([], new_ast), do: new_ast
+
+  def generate_gamma_charge(computation) do
+    computation
+    |> gamma_for_computation
+    |> (&(generate_from_source("UltraDark.Contract.charge_gamma(#{&1})").body)).()
+    |> List.first
+  end
+
+  def gamma_for_computation(%ESTree.BinaryExpression{ operator: operator }), do: compute_gamma_for_operator(operator)
+  def gamma_for_computation(%ESTree.UpdateExpression{ operator: operator }), do: compute_gamma_for_operator(operator)
+  def gamma_for_computation(%ESTree.ExpressionStatement{ expression: expression }), do: gamma_for_computation(expression)
+  def gamma_for_computation(%ESTree.ReturnStatement{ argument: argument }), do: gamma_for_computation(argument)
+  def gamma_for_computation(%ESTree.VariableDeclaration{ declarations: declarations }), do: gamma_for_computation(declarations)
+  def gamma_for_computation(%ESTree.VariableDeclarator{ init: %{ value: value } }), do: calculate_gamma_for_declaration(value)
+  # def gamma_for_computation(%ESTree.AssignmentExpression{ left: left }), do: IEx.pry
+  def gamma_for_computation(%ESTree.CallExpression{}), do: 0
+
+  def gamma_for_computation([first | rest]), do: gamma_for_computation(rest, [gamma_for_computation(first)])
+  def gamma_for_computation([first | rest], gamma_list), do: gamma_for_computation(rest, [gamma_for_computation(first) | gamma_list])
+  def gamma_for_computation([], gamma_list), do: Enum.reduce(gamma_list, fn gamma, acc -> acc + gamma end)
+
+  def gamma_for_computation(other) do
+    IO.warn "Gamma for computation not implemented for: #{other.type}"
+    # IEx.pry
+    0
+  end
+
+  @doc """
+    Takes in a variable declaration and returns the gamma necessary to store the data
+    within the contract. The cost is mapped to 2500 gamma per byte
+  """
+  @spec calculate_gamma_for_declaration(any) :: number
+  def calculate_gamma_for_declaration(value) do
+    (value |> :erlang.term_to_binary |> byte_size) * 2500 # Is there a cleaner way to calculate the memory size of any var?
+  end
+
+  @doc """
+    Takes in a binary tree expression and returns the amount of gamma necessary
+    in order to perform the expression
+  """
+  @spec compute_gamma_for_operator(atom) :: number | {:error, String.t}
+  defp compute_gamma_for_operator(operator) do
```

Comment:
@alexdovzhanyn this could be something around these lines

```
defp compute_gamma_for_operator(operator) when op in @base, do: 2
defp compute_gamma_for_operator(operator) when op in @low, do: 3
defp compute_gamma_for_operator(operator), do: {:error, "No compute_binary_or_update_expression_gamma defined for operator: #{op}"}
```

Reply:
@alexdovzhanyn also I would suggest to use atoms for represent errors, if at any given time you need to show the human readable error you can always convert it for example

```elixir
{:error,  { :no_compute_binary_or_update_expression_gamma, op }}
```

```elixir
defmodule ErrorTranslator do

  def humanize( {:error,  { :no_compute_binary_or_update_expression_gamma, op }} ) do
    "No compute_binary_or_update_expression_gamma defined for operator: #{op}"}
  end
end
```

Something like that,

I recommend you to read http://michal.muskala.eu/2017/02/10/error-handling-in-elixir-libraries.html